smart-contracts/contracts/cl-vault/src/helpers.rs

```diff
@@ -290,139 +326,185 @@ pub fn get_max_utilization_for_ratio(
+    let ratios = ps.into_iter().map(|(ps, ps_ratio)| {
+        let r0 = Decimal::from_ratio(ps_ratio.asset0, total0);
+        let r1 = Decimal::from_ratio(ps_ratio.asset1, total1);
+        let ratio = min(r0, r1);
+
+        (ps, ratio)
+    });
+
+    let total_ratio = ratios
+        .clone()
+        .fold(Decimal::zero(), |acc, (_, ratio)| acc + ratio);
+
+    let psf: Result<Vec<(Position, Uint128, Uint128)>, ContractError> = ratios
+        .into_iter()
+        .map(|(ps, ratio)| {
+            let normalized = ratio / total_ratio;
+
+            Ok((ps, token0 * normalized, token1 * normalized))
+        })
+        .collect();
+    psf
+}
+
+/// per position, calculate the ratio of tokens that position needs, multiplied with the ratio of that position
+/// within the vault
+fn get_min_ratio_per_position(
+    positions: Vec<(Position, FullPositionBreakdown)>,
+    spot_price: Decimal,
+) -> Result<Vec<(Position, PositionRatio)>, ContractError> {
+    // per position, calculate the ratio of asset1 and asset2 a position needs
+
+    // for each position, we take some set amount of liquidity and calculate how many asset0 and asset1 those would give us
+    // we assume
+    let liquidity: Decimal256 = Decimal::new(100_000_000_000_000_000_000_000_u128.into()).into();
+    let ps: Result<Vec<(Position, PositionRatio)>, ContractError> = positions
+        .into_iter()
+        .map(|(p, fp)| {
+            let pos = fp.position.unwrap();
+            let upper_price_sqrt = tick_to_price(pos.upper_tick)?.sqrt();
+            let lower_price_sqrt = tick_to_price(pos.lower_tick)?.sqrt();
+            let spot_price_sqrt: Decimal256 = spot_price.sqrt().into();
+
+            let amount0 =
+                asset0(liquidity, spot_price_sqrt, upper_price_sqrt).unwrap_or(Uint128::zero());
+            let amount1 =
+                asset1(liquidity, lower_price_sqrt, spot_price_sqrt).unwrap_or(Uint128::zero());
+            Ok((p, PositionRatio::new(amount0, amount1)))
+        })
+        .collect();
+
+    let positions = ps?;
+
+    let total_ratio = positions
+        .iter()
+        .fold(Uint128::zero(), |acc, (p, _)| acc + p.ratio);
+
+    // now that we know how much tokens a positions internal ratio needs, we need to normalize these internal ratios to eachother using the positions ratios
+    // Each position might get ratio/total_ratio of tokens. We want to find the effective ratio for each position then.
+    // we multiply the position's internal ratio by the positions external ratio
+    // The external ratio is the positions ratio divided by the total ratio of all positions
+    let positions: Result<Vec<(Position, PositionRatio)>, ContractError> = positions
+        .into_iter()
+        .map(|(p, internal_ratio)| {
+            let external_ratio = Decimal::from_ratio(p.ratio, total_ratio);
+            let total_ratio = internal_ratio.checked_mul_ratio(external_ratio)?;
```

Comment:
Should we call this effective ratio after multiplying internal and external to distinguish it with previously calculated total_ratio.

Reply:
Good one

Reply:
changed in [bc4925c](https://github.com/quasar-finance/quasar/commit/bc4925c356621b7777e9fafdaec620baab63f86a)